### ISSUE TITLE
Add a daterange element subplugin

### ISSUE THREAD

**Issue by abias** [markn86/moodle-mod_customcert#185]
In some scenarios, you might need to create a certificate about a given lecture term, when it's not important at which day it was issues but in which term it was issued.

We propose to add a new datarange element subplugin, based on the date element subplugin, which automatically shows the configured daterange string from the configures daterange.

For example:
* The teacher adds the daterange element to his template
* He configures the daterange string to be displayed:
** From 1st april to 30th september: String is "summer term"
** From 1st october to 30th march: String is "winter term"
* The teacher can still use the "date item" setting where he can choose if the issued date, the course start date and so on should be used.
* Assumed that date item is set to "issued date": If the student downloads the certificate at 5th may 2018, the daterange element will output "summer term 2018".

**Comment by dmitriim** [comment]
@abias Would you think these data range strings should be flexible and you can have as many as you'd like per one element?

**Comment by abias** [comment]
@dmitriim: For our particular use case, two fixed strings will be enough (as we want to use the date range strings for labeling lecture terms). But I am sure that this won't be sufficient for all users. At least, there are institutions out there working with trimesters or quarters. And there might be users who will want to use the date range for something completely different like labeling calendar weeks.

So, I would say:
* If we will go with a fixed amount of date range strings, I would propose to have at least 10 strings available.
* If it will be easy to implement with a small overhead (and I fear that this won't be easy as Moodle core does not offer an API function as far as I remember), I would propose to start with 2 strings and to let the user add more strings as needed.

**Comment by dmitriim** [comment]
Pull request for this feature https://github.com/markn86/moodle-mod_customcert/pull/216

**Comment by abias** [comment]
Hi @dmitriim,

I have just tested your patch from #216 for this issue (together with the patch from #213) on a local test instance.

Here a some minor things to fix quickly:
* When creating a daterange element (on http://localhost:8040/mod/customcert/edit_element.php?tid=1&action=add&element=daterange&pageid=1) and when saving the daterange element, I got these debug messages:
```
Did you remember to call setType() for 'numranges'? Defaulting to PARAM_RAW cleaning.
line 1354 of /lib/formslib.php: call to debugging()
line 312 of /lib/formslib.php: call to moodleform->detectMissingSetType()
line 218 of /lib/formslib.php: call to moodleform->_process_submission()
line 79 of /mod/customcert/edit_element.php: call to moodleform->__construct()
Did you remember to call setType() for 'datestring0'? Defaulting to PARAM_RAW cleaning.
line 1354 of /lib/formslib.php: call to debugging()
line 312 of /lib/formslib.php: call to moodleform->detectMissingSetType()
line 218 of /lib/formslib.php: call to moodleform->_process_submission()
line 79 of /mod/customcert/edit_element.php: call to moodleform->__construct()
Did you remember to call setType() for 'datestring1'? Defaulting to PARAM_RAW cleaning.
line 1354 of /lib/formslib.php: call to debugging()
line 312 of /lib/formslib.php: call to moodleform->detectMissingSetType()
line 218 of /lib/formslib.php: call to moodleform->_process_submission()
line 79 of /mod/customcert/edit_element.php: call to moodleform->__construct()
```
* I would propose to place the "date item" setting between the "element name" and "font" setting similar to the existing date element subplugin.

After testing the functionality of the patch, I am afraid that there seems to be a gap in my specification above which now results in some "friction loss" with the developed code:
* My initial motivation in #185 was to be able to output a text which shows the term name together with the current year, e.g. "summer term 2018" and "winter term 2018". For our needs, it would have been great to set the cut-off dates between the date ranges in the plugin without a specific year. But the specification lacks this fact.
* The current implementation now has a datepicker (day / month / year) for each daterange. I am able to define cut-off days (including a specific year) and the PDF preview shows the mapped date range string.
* Now here are the two problems which we would have to discuss:
** If the date to be mapped does not match any configured date range, the PDF preview shows the current datetime with some local formatting. This is not ideal and should be considered as bug. I think if the date cannot be mapped, there shouldn't be any output at all. If possible, a new setting for a (optional) fallback string would be great.
** I admit that there might be usage scenarios where there are different cut-off dates each year and the current implementation is fine for them. For our particular usage scenario where the cut-off dates are the same each year, the current implementation would mean that we would have to add enough dateranges for the next years when the certificate is configured. However, I would like to think about possible improvements:
If possible, I would propose that the dateranges can be set _with_ a specific year or _without_ a specific. Adding a setting for this decision to element/daterange/settings.php shouldn't be a problem, but as far as I see there is no date_selector without years in Moodle forms.
Additionally, we would then need a placeholder which can be added to each daterange string and which will be replaced in the PDF with the specific year.
